_posts/2023-10-02-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,457 @@
+
+You might also want to explore how to expand and improve on other aspects of your client through configuration.
+
+**Implementing security**
+
+Ensure a secure connection when connecting to your Kafka cluster by implementing security measures for authentication, encryption, and authorization. In Strimzi, this process involves configuring listeners and user accounts.
+
+If you're not familiar with the process of implementing security, we summarized it in the post for the [example producer client](https://strimzi.io/blog/2023/09/25/kafka-producer-client-essentials/).
+
+**Avoiding data loss or duplication** 
+
+As mentioned, we use `earliest` as our offset reset policy to avoid the potential for data loss by sticking with the `latest` default.
+
+We also disable the default Kafka capability to commit offsets automatically.
+The Kafka auto-commit mechanism is convenient but introduces a risk of data loss and duplication. 
+To disable the auto-commit mechanism, we set `ENABLE_AUTO_COMMIT_CONFIG` to `false`. 
+Instead, we use a combination of the `commitAsync` and `commitSync` APIs to manage offset commits within the application.
+
+We use `commitAsync` for regular offset commits, allowing the consumer to continue polling for new messages without waiting for the offsets to be committed.
+
+However, if the `OffsetOutOfRangeException` or `NoOffsetForPartitionException` errors are thrown, or if we encounter an unrecoverable error, we use `commitSync` to ensure that offsets are correctly handled and committed. 
+By using the `commitSync` API, the consumer application will not poll for new messages until the last offset in the batch is committed. 
+
+**Recovering from failure**
+
+Kafka provides mechanisms to detect and recover from failures within consumer groups. Two useful properties that aid in this recovery process are `session.timeout.ms` and `heartbeat.interval.ms`.
+
+The `session.timeout.ms` property specifies the maximum amount of time a consumer in a group can go without sending a _heartbeat_ to the Kafka broker. 
+This heartbeat is an indicator that the consumer is active in the group. 
+If a consumer fails to send a heartbeat within the specified session timeout, Kafka considers it as a failure, and the consumer is marked as inactive.
+Setting this value too low can result in false-positive outcomes, while setting it too high can lead to delayed recovery from failures.
```

Comment:
maybe it's worth mentioning that a consumer marked as inactive also starts a rebalancing of the partitions for the topic.